packages/common/src/hardhat/tasks/artifacts.js

```diff
@@ -0,0 +1,206 @@
+const path = require("path");
+const fs = require("fs");
+const uniqBy = require("lodash.uniqby");
+
+const { task, types } = require("hardhat/config");
+
+function removeFileIfExists(filename) {
+  try {
+    fs.unlinkSync(filename);
+  } catch (e) {
+    // do nothing.
+  }
+}
+
+async function getArtifactPathList(hre, relativeTo) {
+  const artifactPaths = await hre.artifacts.getArtifactPaths();
+
+  // Generate a unique list of artifacts and paths to them. Unique is necessary because there are some redundantly
+  // named contracts.
+  return uniqBy(
+    artifactPaths.map((artifactPath) => ({
+      contractName: path.basename(artifactPath).split(".")[0],
+      relativePath: `./${path.relative(path.dirname(relativeTo), artifactPath)}`,
+    })),
+    "contractName"
+  );
+}
+
+function getCorePath(hre, relativeTo) {
+  const artifactPath = hre.config.paths.artifacts ? path.join(hre.config.paths.artifacts, "../") : "./";
+  return `./${path.relative(relativeTo, artifactPath)}`;
+}
+
+function getAddressesMap(hre) {
+  // Generate a map of name => chain id => address.
```

Comment:
how does this work if there are multiple named deployments for one contract. i.e multiple LSP factories, all equally "valid"? based off this I'm guessing it just has the latest but in general how would we expand this to accommodate more flexibility in this regard?

Reply:
If you intend to have multiple deployments of a contract, you can use the new field "deploymentName" in the config to denote the name of the deployment if it is different from the contract name (it defaults to contractName if deploymentName isn't there). Then, that address is accessed using that name rather than the contract name. That's why I elected to use the more general term "name" in the getAddress function.

Reply:
makes sence!